diff -urN xfce4-notes-plugin-1.6.0/configure.in xfce4-notes-plugin-1.6.0.new/configure.in
--- xfce4-notes-plugin-1.6.0/configure.in	2007-11-23 00:39:37.000000000 +0100
+++ xfce4-notes-plugin-1.6.0.new/configure.in	2007-12-07 18:21:14.000000000 +0100
@@ -52,7 +52,7 @@
 XDT_CHECK_PACKAGE([LIBXFCE4PANEL], [libxfce4panel-1.0], [4.3.90])
 
 dnl Translations
-XDT_I18N([ca cs de en_GB eu fr hu nb_NO nl pl pt_BR pt_PT zh_TW])
+XDT_I18N([ca cs de en_GB eu fr hu nb nl pl pt_BR pt zh_TW])
 
 dnl Check for debugging support
 XDT_FEATURE_DEBUG()
diff -urN xfce4-notes-plugin-1.6.0/po/LINGUAS xfce4-notes-plugin-1.6.0.new/po/LINGUAS
--- xfce4-notes-plugin-1.6.0/po/LINGUAS	2007-11-23 00:35:01.000000000 +0100
+++ xfce4-notes-plugin-1.6.0.new/po/LINGUAS	2007-12-07 18:21:00.000000000 +0100
@@ -1,2 +1,2 @@
 # set of available languages (in alphabetic order)
-ca cs de en_GB eu fr hu nb_NO nl pl pt_BR pt_PT zh_TW
+ca cs de en_GB eu fr hu nb nl pl pt_BR pt zh_TW
